GenAPI: add support for reference parameters in constructors.

Repro steps:
1. `generate.sh --package "System.Threading,4.0.11"`
2. `dotnet build src/referencePackages/System.Threading/4.0.11/System.Threading.4.0.11.csproj`

Link to failing build: https://dev.azure.com/dnceng-public/public/_build/results?buildId=194555&view=logs&jobId=2f0d093c-1064-5c86-fc5b-b7b1eca8e66a&j=2f0d093c-1064-5c86-fc5b-b7b1eca8e66a&t=d2b639a6-cb19-5f1f-66fd-8047f66b3026

Failures:
```
  CSC : error CS1616: Option 'DelaySign' overrides attribute 'System.Reflection.AssemblyDelaySignAttribute' given in a source file or added module [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.0] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  CSC : error CS1616: Option 'PublicSign' overrides attribute 'System.Reflection.AssemblyDelaySignAttribute' given in a source file or added module [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.0] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  CSC : error CS1616: Option 'CryptoKeyFile' overrides attribute 'System.Reflection.AssemblyKeyFileAttribute' given in a source file or added module [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.0] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  CSC : error CS8105: Attribute 'System.Reflection.AssemblyKeyFileAttribute' is ignored when public signing is specified. [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.0] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  /__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/ref/netstandard1.0/System.Threading.cs(149,16): error CS0177: The out parameter 'createdNew' must be assigned to before control leaves the current method [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.0] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  /__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/ref/netstandard1.0/System.Threading.cs(318,16): error CS0177: The out parameter 'createdNew' must be assigned to before control leaves the current method [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.0] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  /__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/ref/netstandard1.0/System.Threading.cs(388,16): error CS0177: The out parameter 'createdNew' must be assigned to before control leaves the current method [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.0] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  /__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/ref/netstandard1.3/System.Threading.cs(169,16): error CS0177: The out parameter 'createdNew' must be assigned to before control leaves the current method [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.3] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  /__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/ref/netstandard1.3/System.Threading.cs(345,16): error CS0177: The out parameter 'createdNew' must be assigned to before control leaves the current method [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.3] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  /__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/ref/netstandard1.3/System.Threading.cs(415,16): error CS0177: The out parameter 'createdNew' must be assigned to before control leaves the current method [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.3] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  CSC : error CS1616: Option 'DelaySign' overrides attribute 'System.Reflection.AssemblyDelaySignAttribute' given in a source file or added module [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.0] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  CSC : error CS1616: Option 'PublicSign' overrides attribute 'System.Reflection.AssemblyDelaySignAttribute' given in a source file or added module [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.0] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  CSC : error CS1616: Option 'CryptoKeyFile' overrides attribute 'System.Reflection.AssemblyKeyFileAttribute' given in a source file or added module [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.0] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  CSC : error CS8105: Attribute 'System.Reflection.AssemblyKeyFileAttribute' is ignored when public signing is specified. [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.0] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  /__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/ref/netstandard1.0/System.Threading.cs(149,16): error CS0177: The out parameter 'createdNew' must be assigned to before control leaves the current method [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.0] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  /__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/ref/netstandard1.0/System.Threading.cs(318,16): error CS0177: The out parameter 'createdNew' must be assigned to before control leaves the current method [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.0] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  /__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/ref/netstandard1.0/System.Threading.cs(388,16): error CS0177: The out parameter 'createdNew' must be assigned to before control leaves the current method [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.0] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  /__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/ref/netstandard1.3/System.Threading.cs(169,16): error CS0177: The out parameter 'createdNew' must be assigned to before control leaves the current method [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.3] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  /__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/ref/netstandard1.3/System.Threading.cs(345,16): error CS0177: The out parameter 'createdNew' must be assigned to before control leaves the current method [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.3] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  /__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/ref/netstandard1.3/System.Threading.cs(415,16): error CS0177: The out parameter 'createdNew' must be assigned to before control leaves the current method [/__w/1/s/artifacts/source-build/self/src/src/referencePackages/src/system.threading/4.0.11/System.Threading.4.0.11.csproj::TargetFramework=netstandard1.3] [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
  /__w/1/s/artifacts/source-build/self/src/eng/SourceBuild.props(18,5): error MSB3073: The command "./build.sh --configuration Release /bl:/__w/1/s/artifacts/source-build/self/src/artifacts/sourcebuild-dependency-projects.binlog /p:LocalNuGetPackageCacheDirectory=/__w/1/s/artifacts/source-build/self/src/artifacts/obj/source-built-upstream-cache/ /p:BuildDependencyPackageProjects=true /p:SetUpSourceBuildIntermediateNupkgCache=true /p:DotNetBuildFromSource=true /p:ArcadeInnerBuildFromSource=true" exited with code 1. [/__w/1/s/.packages/microsoft.dotnet.arcade.sdk/8.0.0-beta.23153.1/tools/Build.proj]
      0 Warning(s)
      21 Error(s)
```

> error CS0177: The out parameter 'createdNew' must be assigned to before control leaves the current method 

We have a logic in the GenAPI to add throw an exception body to methods with non-void return type or at least on out parameter.  Constructors may have `out` parameters too:

```C#
public partial class EventWaitHandle : WaitHandle
{
  public EventWaitHandle(bool initialState, EventResetMode mode, string name, out bool createdNew) { }
}
```

> error CS1616: Option 'DelaySign' overrides attribute 'System.Reflection.AssemblyDelaySignAttribute' given in a source file or added module

We can add such attributes to an exclusion list of attributes of the SBRP generator; not an issue in the GenAPI

--- a/src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxRewriter/BodyBlockCSharpSyntaxRewriter.cs
+++ b/src/GenAPI/Microsoft.DotNet.GenAPI/SyntaxRewriter/BodyBlockCSharpSyntaxRewriter.cs
@@ -25,9 +25,21 @@ namespace Microsoft.DotNet.GenAPI.SyntaxRewriter
         /// <inheritdoc />
         public override SyntaxNode? VisitConstructorDeclaration(ConstructorDeclarationSyntax node)
         {
-            node = node.WithBody(GetEmptyBody())
-                       .WithParameterList(node.ParameterList.WithTrailingTrivia(SyntaxFactory.Space));
-            return base.VisitConstructorDeclaration(node);
+            // visit subtree first to normalize type names.
+            ConstructorDeclarationSyntax? rs = (ConstructorDeclarationSyntax?)base.VisitConstructorDeclaration(node);
+            if (rs is null) return rs;
+
+            // if there is at least one reference parameter - generate non empty body.
+            if (node.ParameterList.Parameters.Any(p => p.Modifiers.Any(m => m.IsKind(SyntaxKind.OutKeyword))))
+            {
+                rs = rs.WithBody(GetThrowNullBody());
+            }
+            else
+            {
+                rs = rs.WithBody(GetEmptyBody());
+            }
+
+            return rs;
         }
 
         /// <inheritdoc />
@@ -47,8 +59,7 @@ namespace Microsoft.DotNet.GenAPI.SyntaxRewriter
                 rs = rs.WithExpressionBody(null);
             }
 
-            string returnType = rs.ReturnType.ToString();
-            if ((returnType != "void" && returnType != "System.Void") ||
+            if (!(rs.ReturnType is PredefinedTypeSyntax predefined && predefined.Keyword.IsKind(SyntaxKind.VoidKeyword)) ||
                 // if there is at least one reference parameter - generate non empty body.
                 node.ParameterList.Parameters.Any(p => p.Modifiers.Any(m => m.IsKind(SyntaxKind.OutKeyword))))
             {

--- a/src/Tests/Microsoft.DotNet.GenAPI.Tests/SyntaxRewriterTests.cs
+++ b/src/Tests/Microsoft.DotNet.GenAPI.Tests/SyntaxRewriterTests.cs
@@ -462,6 +462,30 @@ namespace Microsoft.DotNet.GenAPI.Tests
                 }
                 """);
         }
+
+        [Fact]
+        public void TestConstructorWithReferenceParameters()
+        {
+            CompareSyntaxTree(new BodyBlockCSharpSyntaxRewriter("Not implemented"),
+                original: """
+                namespace A
+                {
+                    public class Foo
+                    {
+                        public Foo(int a, out int b) { b = 1; }
+                    }
+                }
+                """,
+                expected: """
+                namespace A
+                {
+                    public class Foo
+                    {
+                        public Foo(int a, out int b) { throw new PlatformNotSupportedException("Not implemented"); }
+                    }
+                }
+                """);
+        }
     }
 
     public class TypeForwardAttributeCSharpSyntaxRewriterTests : SyntaxRewriterTests

--- a/src/Tests/Microsoft.DotNet.GenAPI.Tests/SyntaxRewriterTests.cs
+++ b/src/Tests/Microsoft.DotNet.GenAPI.Tests/SyntaxRewriterTests.cs
@@ -472,6 +472,7 @@ namespace Microsoft.DotNet.GenAPI.Tests
                 {
                     public class Foo
                     {
+                        public Foo(int a) { }
                         public Foo(int a, out int b) { b = 1; }
                     }
                 }
@@ -481,6 +482,7 @@ namespace Microsoft.DotNet.GenAPI.Tests
                 {
                     public class Foo
                     {
+                        public Foo(int a) { }
                         public Foo(int a, out int b) { throw new PlatformNotSupportedException("Not implemented"); }
                     }
                 }

